src/main/java/datawave/query/util/AllFieldMetadataHelper.java

```diff
@@ -1050,6 +1058,348 @@ public Set<String> loadDatatypes() throws TableNotFoundException {
+     */
+    private Map<String,Map<String,FieldIndexHole>> getFieldIndexHoles(Text indexColumnFamily) throws TableNotFoundException {
+        log.debug("cache fault for getFieldIndexHoles(" + this.auths + "," + this.metadataTableName + ")");
+        
+        Scanner bs = ScannerHelper.createScanner(accumuloClient, metadataTableName, auths);
+        
+        // Fetch the frequency column and the specified index column.
+        bs.fetchColumnFamily(ColumnFamilyConstants.COLF_F);
+        bs.fetchColumnFamily(indexColumnFamily);
+        
+        // For all keys in the DatawaveMetadata table.
+        bs.setRange(new Range());
+        
+        // We must first scan over the fieldName-datatype combinations and extract the date ranges in which we've seen them. Each date range represents a span
+        // of time when we saw an event for each day in that date range, from the start (inclusive) to end (inclusive).
+        Map<String,SortedSet<Pair<Date,Date>>> frequencyMap = new HashMap<>();
+        Map<String,SortedSet<Pair<Date,Date>>> indexMap = new HashMap<>();
+        Calendar calendar = Calendar.getInstance();
+        
+        String prevFieldName = null;
+        String prevDatatype = null;
+        Date prevDate = null;
+        Date startDate = null;
+        Text prevColumnFamily = null;
+        
+        // Points to the target map object that we add date ranges to. This changes when we see a different column family compared to the previous row. We must
+        // initially start adding entries to the frequency map.
+        Map<String,SortedSet<Pair<Date,Date>>> dateMap = frequencyMap;
+        
+        Map<String,Multimap<String,Pair<Date,Date>>> fieldIndexHoles = new HashMap<>();
+        
+        // Scan each row and extract the date ranges.
+        for (Entry<Key,Value> entry : bs) {
+            Key key = entry.getKey();
+            String fieldName = key.getRow().toString();
+            Text columnFamily = key.getColumnFamily();
+            
+            // Parse the data type and event date from the column qualifier.
+            String cq = key.getColumnQualifier().toString();
+            int offset = cq.indexOf(NULL_BYTE);
+            String datatype = cq.substring(0, offset);
+            Date date = DateHelper.parse(cq.substring((offset + 1)));
+            
+            // If this is the very first entry we've seen, update the tracking variables and continue to the next entry.
+            if (prevFieldName == null) {
+                prevFieldName = fieldName;
+                prevDatatype = datatype;
+                prevDate = date;
+                startDate = date;
+                prevColumnFamily = columnFamily;
+                continue;
+            }
+            
+            // If the column family is different, determine record the last date range, and begin collecting date ranges for the next batch of related rows.
+            if (!prevColumnFamily.equals(columnFamily)) {
```

Comment:
I think you may need to verify that you are in the same fieldname.  I could see several field names in a row only with F fields and not being indexed.  I think this whole loop might be cleaner if you simply gathered all the data for a fields, then when you get to the next field (or the end) you update the field index wholes.

Reply:
This check is performed [here](https://github.com/NationalSecurityAgency/datawave-metadata-utils/pull/29/files/f2bc5377d7548f9f18283852ff42727a398efd53#diff-940dceae6ce6aaacc9b6c6f5ff78b5508669273ae516eff29ab03ec0efa93aebR1164) if we know that the column family has not changed. I will try to make the documentation surrounding that check a little clearer.

Reply:
I have updated the documentation to improve the clarity.